### PR TITLE
Make token provider scoped per instance

### PR DESCRIPTION
**Pull request by luismfonseca**
We want to scope per instance so that we support multiple instances with different users set to each one.

### CODE REVIEW

File: pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/FakeErrol.kt [hunk 1]
```diff
@@ -49,19 +53,19 @@ class FakeErrol(port: Int, private val clusterKey: String = ""): NanoHTTPDRouter
     post("/instances/{instanceId}/devices/fcm") {
       entity(RegisterDeviceRequest::class) { registerDeviceRequest ->
         val device = FakeErrolDevice.New(registerDeviceRequest.token, mutableSetOf())
-        storage.devices[device.id] = device
+        getInstanceStorage(params["instanceId"]!!).devices[device.id] = device
```
Comment (daniellevass) [inline pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/FakeErrol.kt]: "instanceId" should be a const to be used by all the tests.

File: pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    var setOnSubscriptionsChangedListenerCalledCount2 = 0
+    var lastSetOnSubscriptionsChangedListenerCalledWithInterests2: Set<String>? = null
+    pni2.setOnDeviceInterestsChangedListener(object : SubscriptionsChangedListener {
+      override fun onSubscriptionsChanged(interests: Set<String>) {
+        setOnSubscriptionsChangedListenerCalledCount2++
+        lastSetOnSubscriptionsChangedListenerCalledWithInterests2 = interests
+      }
+    })
+
+    // this will cause it to receive the initial interest set of {"hello"}
```
Comment (daniellevass) [inline pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt]: this comment isn't true? it's an initial interest set of {"potatoes"} - and we were using single versions of things so it should just be one potato 🥔

File: pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+    // make sure the callbacks were correctly triggered
+    assertTrue(successWasCalled1)
+    assertFalse(failureWasCalled1)
+    assertTrue(successWasCalled2)
+    assertFalse(failureWasCalled2)
+  }
+
+  @Test
+  fun setOnDeviceInterestsChangedListenerNotAffectTheOther() {
```
Comment (daniellevass) [inline pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt]: setOnDeviceInterestsChangedListenerNotAffectTheOther -> setOnDeviceInterestsChangedListenerDoesNotAffectTheOther ?

File: pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+  val errolClient1 = ErrolAPI(instanceId1, "http://localhost:8080")
+  val errolClient2 = ErrolAPI(instanceId2, "http://localhost:8080")
+
+  fun getStoredDeviceId(instanceId: String): String? {
+    val deviceStateStore = InstanceDeviceStateStore(InstrumentationRegistry.getTargetContext(), instanceId)
+    return deviceStateStore.deviceId
+  }
+
+  companion object {
+    val errol = FakeErrol(8080, "really-long-cluster-secret-key")
```
Comment (daniellevass) [inline pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt]: we should make "really-long-cluster-secret-key" a const, and then we can use it more safely in the tests too.

File: pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/FakeErrol.kt (hunk 1 above)
Comment (luismfonseca) [reply pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/FakeErrol.kt]: This is the server, it makes no assumptions on instance ids. In fact, the tests use different (and multiple) instance ids.

File: pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt (hunk 2 above)
Comment (luismfonseca) [reply pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt]: I think this was just a case of blind copy-pasting

Comment (luismfonseca) [inline pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt]:
```suggestion
```

File: pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt (hunk 3 above)
Comment (luismfonseca) [reply pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/MultiInstanceSupportTest.kt]:
```suggestion
  fun setOnDeviceInterestsChangedListenerDoesNotAffectTheOther() {
```

File: pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/FakeErrol.kt (hunk 1 above)
Comment (daniellevass) [reply pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/FakeErrol.kt]: Oh I just meant the hardcoded string should be a const? we use the same hardcoded string 11 more times in this file.

Comment (luismfonseca) [reply pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/FakeErrol.kt]: Ah, I see. So this string must match what's on the path. I guess you could interpolate in the path, and have it as a constant, but I don't think it really improves things that much in this case.

Comment (luismfonseca) [reply pushnotifications-integration-tests/src/androidTest/java/com/example/pushnotificationsintegrationtests/FakeErrol.kt]: Ah, I see. So this string must match what's on the path. I guess you could interpolate in the path, and have it as a constant, but I don't think it really improves things that much in this case.

**Pull request merged**